src/transformers/models/bert/modeling_bert.py

```diff
@@ -1517,13 +1518,29 @@ def forward(
 
         loss = None
         if labels is not None:
-            if self.num_labels == 1:
-                #  We are doing regression
-                loss_fct = MSELoss()
-                loss = loss_fct(logits.view(-1), labels.view(-1))
+            if self.problem_type is not None:
+                if self.problem_type == "single_column_regression":
+                    loss_fct = MSELoss()
+                    loss = loss_fct(logits.view(-1), labels.view(-1))
+                elif self.problem_type == "multi_column_regression":
+                    loss_fct = MSELoss()
+                    loss = loss_fct(logits.view(-1, self.num_labels), labels)
+                elif self.problem_type in ("binary_classification", "multi_class_classification"):
+                    loss_fct = CrossEntropyLoss()
+                    loss = loss_fct(logits.view(-1, self.num_labels), labels.view(-1))
+                elif self.problem_type in ("multi_label_classification"):
+                    loss_fct = BCEWithLogitsLoss()
+                    loss = loss_fct(logits, labels)
+                else:
+                    raise Exception("Problem type not understood")
             else:
-                loss_fct = CrossEntropyLoss()
-                loss = loss_fct(logits.view(-1, self.num_labels), labels.view(-1))
+                if self.num_labels == 1:
+                    #  We are doing regression
+                    loss_fct = MSELoss()
+                    loss = loss_fct(logits.view(-1), labels.view(-1))
+                else:
+                    loss_fct = CrossEntropyLoss()
+                    loss = loss_fct(logits.view(-1, self.num_labels), labels.view(-1))
```

Comment:
I was thinking that we should set `problem_type` to a default value if it's None:
- one label or float labels -> regression
- int labels of shape [seq_len] -> single_label_classification
- array of labels of shape [seq_len x num_labels] -> multi-label classification

This way we still save in the config how the model was trained.

Reply:
> array of labels of shape [seq_len x num_labels] -> multi-label classification
this can also mean multi-column regression. i can check dtype though: float => regression, else classification. WDYT?

Reply:
Doesn't work as BCEWithLogitsLoss wants the targets as floats. So we have to pick one default between multi-label and regression for that case. I'd go with multi-label personally.

Reply:
fixed